### ISSUE TITLE
Proposal to define privacy-enhanced prefetching and prerendering

### ISSUE THREAD

**Issue by jeremyroman** [WICG/proposals#2]
tl;dr: We'd like to define how to prefetch and prerender content with more privacy-focused direction, and we think we'll need a mechanism for authors to become eligible since we think it is likely to break existing content otherwise.

Proposed solution explainer: https://github.com/jeremyroman/alternate-loading-modes

---

In order to making the experience of loading on the web faster, user agents employ prefetching and prerendering techniques. However, making cookies and other credentials available to the origin server or script may be inconsistent with the privacy objectives of the user or of the referring site.

First, consider the __fetch__ of the resource. User agents would ideally prefetch the content in a way that does not identify the user. For example, the user agent could:

* send a request without credentials (e.g., no `Cookie` or `Authorization` request header)
* establish the connection from a different client IP address (e.g., using a proxy server or virtual private network, if available)
* use a previously fetched response, including one previously fetched by a third party if it can be authenticated

However, because this (intentionally) obscures the user's identity, the response document cannot be personalized for the user. If it is used when the user navigates, the user will notice that they are not logged in (even if they should be), and other surprising behavior. A page designed with this in mind could "upgrade" itself when it loads, by personalizing the page based on data in unpartitioned storage and by fetching personalized content from the server.

Second, consider __prerendering__ the page. User agents would ideally allow HTML parsing, subresource fetching, and script execution in a way that does not identify the user or cause user-visible annoyance. For example, the user agent could:

* apply mitigations as above to subresource and scripted fetches
* deny scripted access to unpartitioned storage, such as cookies and IndexedDB
* deny permission to invoke `window.alert`, autoplay audio, and other APIs inappropriate at this time

In this case, not only is the HTML resource not personalized, but script will observe restrictions that would not ordinarily apply until navigation actually occurs. A page designed with this in mind could tolerate this at prerender time, and "upgrade" itself on navigation by accessing storage or fetching from the network.

Since existing web pages are unlikely to behave well with these restrictions today, and it is impractical for user agents to distinguish such pages, [we propose a lightweight way][alternate-loading-modes] for a page to declare that it is prepared for this and will, if necessary, upgrade itself when it gains access to unpartitioned storage and other privileges.

There has been previous discussion along these lines in [w3c/resource-hints#82](https://github.com/w3c/resource-hints/issues/82#issuecomment-536492276). (It also proposes a new `prenavigate` hint; defining triggers for these loading modes is not yet part of this proposal.)

[alternate-loading-modes]: https://github.com/jeremyroman/alternate-loading-modes

**Comment by jeremyroman** [comment]
@domenic @kinu @yoavweiss FYI

@annevk @youennf You participated in the discussion on w3c/resource-hints#82 (and a previous discussion at TPAC, per the comment). This kind of idea has come up again within as we (Chrome) look at our strategy for making loading faster, and I'd appreciate your thoughts on this direction.

**Comment by annevk** [comment]
The topic of this issue is preloading, but it doesn't seem to cover preloading at all. As I understand it:

* preloading: fetch subresources (and maybe workers?) early
* prefetching: fetch page to potentially navigate next to early
* prerendering: background-navigate to page to potentially navigate next to early

It would be helpful to be rather explicit about this as the actual terms are rather confusing.

As for feedback:

* Perhaps we should not use the existing confusing terminology in new features.
* As discussed during TPAC, having only a server-side opt-in could mean a lot of wasteful fetches and background-navigations. Declaring those issues out-of-scope doesn't seem great.

**Comment by jeremyroman** [comment]
> * Perhaps we should not use the existing confusing terminology in new features.

Yeah, sorry. I've struggled to come up with an umbrella term for prefetching and prerendering. We've used the term "pre*" internally at Google but that's arguably even less descriptive. Would it be better to say "prefetching/prerendering" everywhere, or to introduce another word starting with "pre"?

> * As discussed during TPAC, having only a server-side opt-in could mean a lot of wasteful fetches and background-navigations. Declaring those issues out-of-scope doesn't seem great.

User agents should definitely not start fetching every resource they can think of willy-nilly.

I think part of it *is* out of scope (e.g., cases initiated by browser-specific features, the most obvious of which is interaction with browser chrome, new tab page, etc), but we also want a web platform feature for the referring page to indicate where this would be suggested.

The most obvious approach is a fresh coat of paint on `<link rel=prerender>` from w3c/resource-hints, but we think we there might be cases where pages today don't use resource hints because of the extra work to determine which of their outgoing links are good candidates for prerendering and could benefit from a broader option (a signal that says "my outgoing links to x, y and z are all fine to prerender; you may do so if your heuristics deem it appropriate"). There's definitely a balance here between being wasteful of resources and making prefetching/prerendering possible when it would be useful.

We'll likely have more to say in this space soon -- and I would appreciate any input you're willing to offer either now or then. This piece of it (a way to restrict the cross-origin case specifically, and a way for the server/author to say that's okay) was already complicated enough that I wanted to engage in the standards space sooner rather than later.

**Comment by yoavweiss** [comment]
This was [discussed](https://docs.google.com/document/d/1inejuvPONXPOLKTCcUzOBhPh6QOckMcltnR-E3xyZVQ/edit?hl=en&forcehl=1#heading=h.nh9xf4k3tm9p) at the WebPerfWG's call a few weeks back.

@achristensen07 @rniwa @bdekoz @pmeenan - we're looking for industry support for this work in order to move it under the WICG. Your comments on that front would be helpful.

**Comment by KenjiBaheux** [comment]
We got permission to post the following quotes:

- "We’re interested in speculation rules and privacy preserving prefetch because that improves loading experience without abusing user privacy." - Yahoo Japan
- "[we are] especially interested in URLPattern based Speculation Rules prefetches because that will add more flexibility to choose links to prefetch that users are likely to click on. We’d like to prefetch top stories links on the home page, speculation rule’s new syntax would make it easy." - Nikkei

**Comment by travisleithead** [comment]
Microsoft is very supportive of this work as well. We have internal partner teams who believe this approach offers the best chance of success for speculatively obtaining resources (developers working together with the browser engine to make the use case work reliably).

**Comment by cwilso** [comment]
This has been migrated: https://github.com/WICG/nav-speculation.

**Comment by pavanratnakar** [comment]
Yahoo is interested in URLPattern based Speculation Rules prefetches. Lot of our pages are SSR and we would love to enhance our navigation experience by prefetching links we think users would likely click.
